src/viam/sdk/common/service_helper.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <type_traits>
+
+#include <viam/sdk/resource/resource_server_base.hpp>
+
+namespace viam {
+namespace sdk {
+
+class ServiceHelperBase {
+   public:
+    ::grpc::Status fail(::grpc::StatusCode code, const char* message) const noexcept;
+
+    ::grpc::Status failNoRequest() const noexcept;
+
+    ::grpc::Status failNoResource(const std::string& name) const noexcept;
+
+    ::grpc::Status failStdException(const std::exception& xcp) const noexcept;
+
+    ::grpc::Status failUnknownException() const noexcept;
```

Comment:
These functions could be made `protected` instead since they are really only used in the implementation of `ServiceHelper::operator()`, but I didn't see much need to hide them away.

Reply:
What's the purpose of having both `ServiceHelper` and `ServiceHelperBase` if you're not hiding these methods and there are no other inheritors of `ServiceHelperBase`?

Reply:
Because these functions don't depend on the template parameters like `ServiceType` and `RequestType`. By moving them into a base class, there is just one definition of these functions, rather than one definition per template instantiation. It also means the implementation can be out-of-line.

Reply:
> rather than one definition per template instantiation

As in the compiler would create a new definition per template instantiation? Is that bad?

> It also means the implementation can be out-of-line.

Gotcha.. didn't think about that constraint.

Reply:
> > rather than one definition per template instantiation
> 
> As in the compiler would create a new definition per template instantiation? Is that bad?
> 

It'll cost you at least compile time. All object files will be a little bigger. Maybe bloat in the linked image unless the compiler figures out that it can common them all together. Harder to set breakpoints. More recompile if you ever need to change the definition. Etc.

Reply:
Cool makes sense! Thanks.